cumulusci/core/dependencies/dependencies.py

```diff
@@ -0,0 +1,910 @@
+    namespace: Optional[str]
+    version: Optional[str]
+    version_id: Optional[str]
+    package_name: Optional[str]
+
+    @property
+    def is_resolved(self):
+        return True
+
+    @property
+    def is_flattened(self):
+        return True
+
+    @property
+    def package(self):
+        return self.package_name or self.namespace or "Unknown Package"
+
+    @pydantic.root_validator
+    def validate(cls, values):
+        assert (values.get("namespace") and values.get("version")) or values.get(
+            "version_id"
+        ), "Must specify `namespace` and `version`, or `version_id`"
+
+        assert None in [
+            values.get("namespace"),
+            values.get("version_id"),
+        ], "Must not specify `namespace`/`version` and `version_id`"
+        return values
+
+    def install(
+        self,
+        context: BaseProjectConfig,
+        org: OrgConfig,
+        options: ManagedPackageInstallOptions = None,
+    ):
+        if not options:
+            options = ManagedPackageInstallOptions()
+
+        if self.namespace and self.version:
+            context.logger.info(
+                "Installing {} version {}".format(self.package, self.version)
+            )
+            install_1gp_package_version(
+                context,
+                org,
+                self.namespace,
+                self.version,
+                options,
+                retry_options=DEFAULT_PACKAGE_RETRY_OPTIONS,
+            )
+        elif self.version_id:
+            context.logger.info(f"Installing {self.version_id}")
+            install_package_version(
+                context,
+                org,
+                self.version_id,
+                options,
+                retry_options=DEFAULT_PACKAGE_RETRY_OPTIONS,
+            )
+
+    @property
+    def name(self):
+        return f"Install {self.package} {self.version or self.version_id}"
+
+    def __str__(self):
+        return self.name
+
+
+class UnmanagedDependency(GitHubRepoMixin, StaticDependency):
+    zip_url: Optional[AnyUrl]
+
+    # or
+    repo_owner: Optional[str]
+    repo_name: Optional[str]
+
+    # or
+    github: Optional[AnyUrl]
+
+    # and
+    ref: Optional[str]
+
+    unmanaged: Optional[bool]
+    subfolder: Optional[str]
+    namespace_inject: Optional[str]
+    namespace_strip: Optional[str]
+
+    @pydantic.root_validator
+    def validate(cls, values):
+        if values.get("repo_owner") or values.get("repo_name"):
+            logger.warning(
+                "The repo_name and repo_owner keys are deprecated. Please use the github key."
+            )
+        assert (
+            values.get("zip_url")
+            or (values.get("github") and values.get("ref"))
+            or (
+                values.get("repo_name")
+                and values.get("repo_owner")
+                and values.get("ref")
+            )
+        ), "Must specify `zip_url`, or `github` and `ref`"
```

Comment:
Would this get any clearer if zipurl-based and github-based unpackaged dependencies were processed by separate dependency classes?

Reply:
I think it would. That's a nice clarity improvement. Will change.

Reply:
These have now been broken out into separate classes. I similarly broke out version id-based package dependencies and version/namespace dependencies, which gets to a 1:1 relationship between dependency class and install pathway.